browser/src/statechart.js

```diff
@@ -0,0 +1,615 @@
+};
+
+const swapFrame = () => {
+  const args = {
+    label_1: getModel().foreground,
+    label_2: getModel().background,
+    frame: getModel().frame
+  };
+  const action = new BackendAction('swap_single_frame', args);
+  addFencedAction(action);
+};
+
+const replaceFrame = () => {
+  const args = {
+    label_1: getModel().foreground,
+    label_2: getModel().background,
+    // frame: getModel().frame? may need to add if we no longer store frame on backend
+  };
+  const action = new BackendAction('replace_single', args);
+  addFencedAction(action);
+};
+
+const replaceAll = () => {
+  const args = {
+    label_1: getModel().foreground,
+    label_2: getModel().background
+  };
+  const action = new BackendAction('replace', args);
+  addFencedAction(action);
+};
+
+const watershed = (context) => {
+  const args = {
+    frame: getModel().frame,
+    label: context.storedLabel,
+    x1_location: context.storedX,
+    y1_location: context.storedY,
+    x2_location: getCanvas().imgX,
+    y2_location: getCanvas().imgY
+  };
+  const action = new BackendAction('watershed', args);
+  addFencedAction(action);
+};
+
+// tool states
+const paintState = {
+  initial: 'idle',
+  entry: assign({ trace: [] }),
+  states: {
+    idle: {
+      on: {
+        mousedown: {
+          // prevents select and paint at the same time
+          cond: (_, event) => !event.shiftKey,
+          target: 'dragging',
+          actions: 'addToTrace'
+        }
+      }
+    },
+    dragging: {
+      on: {
+        mousemove: { actions: 'addToTrace' },
+        mouseup: { actions: 'draw' },
+      }
+    }
+  }
+};
+
+const thresholdState = {
+  initial: 'idle',
+  states: {
+    idle: {
+      on: {
+        mousedown: {
+          target: 'dragging',
+          actions: 'storeClick'
+        }
+      }
+    },
+    dragging: {
+      on: {
+        mouseup: [
+          {
+            cond: 'nonemptyBox',
+            actions: 'threshold'
+          },
+          { target: 'idle' },
+        ]
+      }
+    }
+  },
+  on: {
+    TOGGLERGB: {
+      target: 'paint',
+      in: '#deepcellLabel.rgb.oneChannel'
+    },
+  }
+};
+
+const floodState = {
+  on: {
+    click: { actions: 'flood' }
+  }
+};
+
+const trimState = {
+  on: {
+    click: { actions: 'trim' }
+  }
+};
+
+const erodeDilateState = {
+  on: {
+    mousedown: {
+      actions: choose([
+        {
+          cond: 'leftMouse',
+          actions: 'erode'
+        },
+        {
+          cond: 'rightMouse',
+          actions: 'dilate'
+        }
+      ])
+    }
+  }
+};
+
+const autofitState = {
+  on: {
+    click: { actions: 'autofit' },
+    TOGGLERGB: {
+      target: 'paint',
+      in: '#deepcellLabel.rgb.oneChannel'
+    },
+  }
+};
+
+const watershedState = {
+  entry: 'storeClick',
+  initial: 'idle',
+  states: {
+    idle: {
+      on: {
+        click: {
+          cond: 'validSeed',
+          target: 'clicked',
+          actions: 'storeClick'
+        }
+      }
+    },
```

Comment:
What does the `entry` field do? Is it storing the last click even if it's in the `idle` state?

Reply:
Good catch, that was from a previous workflow with watershed where we entered it on click instead of on a keybind. It would store the coordinates and labels below the mouse when entering the watershed state. These will be overwritten once we click in the idle state and call `storeClick` again, so this doesn't cause any bugs, but we should remove it.